pkg/proc/disasm_arm64.go

```diff
@@ -0,0 +1,163 @@
+// the flavour specified.
+func (inst *AsmInstruction) Text(flavour AssemblyFlavour, bi *BinaryInfo) string {
+	if inst.Inst == nil {
+		return "?"
+	}
+
+	var text string
+
+	switch flavour {
+	case GNUFlavour:
+		text = arm64asm.GNUSyntax(arm64asm.Inst(*inst.Inst))
+	default:
+		text = arm64asm.GoSyntax(arm64asm.Inst(*inst.Inst), inst.Loc.PC, bi.symLookup, nil)
+	}
+
+	return text
+}
+
+// IsCall returns true if the instruction is a BL or BLR instruction.
+func (inst *AsmInstruction) IsCall() bool {
+	if inst.Inst == nil {
+		return false
+	}
+	return inst.Inst.Op == arm64asm.BL || inst.Inst.Op == arm64asm.BLR
+}
+
+// IsRet returns true if the instruction is a RET or ERET instruction.
+func (inst *AsmInstruction) IsRet() bool {
+	if inst.Inst == nil {
+		return false
+	}
+	return inst.Inst.Op == arm64asm.RET || inst.Inst.Op == arm64asm.ERET
+}
+
+func resolveCallArg(inst *archInst, instAddr uint64, currentGoroutine bool, regs Registers, mem MemoryReadWriter, bininfo *BinaryInfo) *Location {
+	if inst.Op != arm64asm.BL && inst.Op != arm64asm.BLR {
+		return nil
+	}
+
+	var pc uint64
+	var err error
+
+	switch arg := inst.Args[0].(type) {
+	case arm64asm.Imm:
+		pc = uint64(arg.Imm)
+	case arm64asm.Reg:
+		if !currentGoroutine || regs == nil {
+			return nil
+		}
+		pc, err = regs.Get(int(arg))
+		if err != nil {
+			return nil
+		}
+	case arm64asm.PCRel:
+		pc = uint64(instAddr) + uint64(arg)
+	default:
+		return nil
+	}
+
+	file, line, fn := bininfo.PCToLine(pc)
+	if fn == nil {
+		return &Location{PC: pc}
+	}
+	return &Location{PC: pc, File: file, Line: line, Fn: fn}
+}
+
+type instrseq []arm64asm.Op
+
+// Possible stacksplit prologues are inserted by stacksplit in
+// $GOROOT/src/cmd/internal/obj/arm64/obj7.go.
+var prologues []instrseq
+
+func init() {
+	var tinyStacksplit = instrseq{arm64asm.MOV, arm64asm.CMP, arm64asm.B}
+	var smallStacksplit = instrseq{arm64asm.SUB, arm64asm.CMP, arm64asm.B}
+	var bigStacksplit = instrseq{arm64asm.CMP, arm64asm.B, arm64asm.ADD, arm64asm.SUB, arm64asm.MOV, arm64asm.CMP, arm64asm.B}
+	var unixGetG = instrseq{arm64asm.LDR}
+
+	prologues = make([]instrseq, 0, 3)
+	for _, getG := range []instrseq{unixGetG} {
+		for _, stacksplit := range []instrseq{tinyStacksplit, smallStacksplit, bigStacksplit} {
+			prologue := make(instrseq, 0, len(getG)+len(stacksplit))
+			prologue = append(prologue, getG...)
+			prologue = append(prologue, stacksplit...)
+			prologues = append(prologues, prologue)
+		}
+	}
+}
+
+// firstPCAfterPrologueDisassembly returns the address of the first
+// instruction after the prologue for function fn by disassembling fn and
+// matching the instructions against known split-stack prologue patterns.
+// If sameline is set firstPCAfterPrologueDisassembly will always return an
+// address associated with the same line as fn.Entry
+func firstPCAfterPrologueDisassembly(p Process, fn *Function, sameline bool) (uint64, error) {
```

Comment:
I think much of this routine (and `checkPrologue`) is arch independent aside from the actual assembly that's being verified. I think if possible let's share this code instead of duplicating it.

Reply:
modified.